### ISSUE TITLE
Corruption when frame size changes

### ISSUE THREAD

**Issue by mcemilg** [cedricve/raspicam#21]
I'm trying to use raspicam_cv to get frames from the camera continuously. I can get the frames properly without making any changes. But if I change the frame sizes, the frames that come in are corrupted.
    First if i add these settings to camera frames that come in corrupted;
```
cap->set(CV_CAP_PROP_FRAME_WIDTH , 176);
cap->set(CV_CAP_PROP_FRAME_HEIGHT, 144);
```

![raspicam_cv_image1](https://cloud.githubusercontent.com/assets/7115634/21098162/b9cb7890-c05f-11e6-88dc-2f63b72d7397.jpg)

This frame is the size I want, but it's still broken.

**Comment by cedricve** [comment]
Does it work with the default width and height?

**Comment by mcemilg** [comment]
This is what happens if I do not call any set function;

![raspicam_cv_image1](https://cloud.githubusercontent.com/assets/7115634/21098703/625f9e62-c062-11e6-90e2-bed65862f8d4.jpg)

But if i call setters with default values like the folloowing;
```
cap.set(CV_CAP_PROP_FRAME_WIDTH , 2592);
cap.set(CV_CAP_PROP_FRAME_HEIGHT, 1944);
```
I am getting this frame, this is corrupted but not as much as the others;

![raspicam_cv_image3](https://cloud.githubusercontent.com/assets/7115634/21098781/b7b57a30-c062-11e6-9c14-dc43719b26fb.jpg)

**Comment by benrules2** [comment]
Are you repeatedly getting a new frame, or just grabbing one? My live feed looks like this sometimes when the camera has just started, but fixes itself. Maybe a 5s sleep before grabbing the image would help.

**Comment by dennisss** [comment]
I am also getting this issue with the the non-opencv video version.

By adding the line: `Camera.setCaptureSize(1640, 1232);` to `simpletest_raspicam.cpp` the following image is produced:

![raspicam_image](https://cloud.githubusercontent.com/assets/1341104/26532740/a693b9d0-43d7-11e7-99ba-9950dd89367f.jpg)

Hardware is Raspberry Pi 3 and Camera v2. The default resolution works fine, but I'm trying to get a higher resolution. Both picamera and raspivid are able to output this resolution, so I'm not sure what's different with this library.

Any ideas? Thanks!

**Comment by FirstCaptain** [comment]
I am having the exact same issue. Has anyone found out a work around or how to fix it yet?

**Comment by dennisss** [comment]
@FirstCaptain yes, I've found the solution and have a rough patch on my branch (https://github.com/dennisss/raspicam/) implemented for video only right now though.

I've tested it with grayscale on the RPi Camera v2, but I think it should work for RGB and BGR as well. YUV needs more work.

The main issue is that if the width and height is not aligned to 32 x 16 blocks, mmal internally pads each row up to that alignment point. 
So the fix is basically to change the routine that copies the buffer to do it row by row and skip the padding: https://github.com/dennisss/raspicam/blob/master/src/private/private_impl.cpp#L766
